cmd/container-disk-v2alpha/main.go

```diff
@@ -0,0 +1,86 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"time"
+
+	"kubevirt.io/client-go/log"
+)
+
+const (
+	ReadinessProbeFile = "/healthy"
+)
+
+func main() {
+	var err error
+
+	var copyPath string
+	var healthCheck bool
+
+	logger := log.DefaultLogger()
+
+	flag.StringVar(&copyPath, "copy-path", "", "Image target path")
+	flag.BoolVar(&healthCheck, "health-check", false, "Do a health check")
+	flag.Parse()
+
+	if !healthCheck && copyPath == "" {
+		logger.Error("No copy-path provides.")
+		os.Exit(1)
+	}
+
+	if healthCheck {
+		_, err := os.Stat(ReadinessProbeFile)
+		if err != nil {
+			os.Exit(1)
+		} else {
+			os.Exit(0)
+		}
+	}
+
+	err = os.MkdirAll(filepath.Dir(copyPath), os.ModePerm)
+	if err != nil {
+		logger.Reason(err).Errorf("Failed to create disk directory %s.", filepath.Dir(copyPath))
+		os.Exit(1)
+	}
+
+	socket := fmt.Sprintf("%s.%s", copyPath, "sock")
+	ln, err := net.Listen("unix", socket)
+	if err != nil {
+		logger.Reason(err).Error("Failed to create socket.")
+		os.Exit(1)
+	}
+	defer ln.Close()
+
+	f, err := os.Create(ReadinessProbeFile)
+	if err != nil {
+		logger.Reason(err).Errorf("Failed to mark myself as ready.")
+		os.Exit(1)
+	}
+	f.Close()
+
+	go func() {
+		for {
+			_, err := ln.Accept()
+			if err != nil {
+				logger.Reason(err).Error("Unrecoverable error on socket.")
+				if _, err := os.Stat(socket); os.IsNotExist(err) {
+					os.Exit(0)
+				}
+				os.Exit(1)
+			}
+		}
+	}()
+
+	for {
+		time.Sleep(1 * time.Second)
+		if _, err := os.Stat(socket); os.IsNotExist(err) {
```

Comment:
is there any reason to trap signals to avoid shutting down before the socket goes away?

Reply:
Since I don't trap it, do you mean that we may have to trap it? I don't think so. The mounted disk is available until virt-handler is done with the cleanup and unmounts the disk.

Reply:
i guess a better way of saying this is... will there be any issues if the containerDisk containers are terminated before virt-launcher?

Reply:
Worst thing is that the kubelet already tries to unmount the containerDisk root, and that you see an error that this failed. As soon as the vmi is down, virt-handler will unmount all containerDisk mounts and the kubelet can proceed with the mount point cleanup.

Reply:
The umount looks like it depends on the isolation pid detection though. It's unclear to me how the umount works after the fact if the pid inside the container exits.

Reply:
Luckily not. It works like this:

 * Every `compute` pod has a container-disk directory mounted which is on the host : `/var/run/kubevirt/container-disks/<uid>` and inside the pod: `/var/run/kubevirt/container-disks/`
 * This means that the pods can only see their disks (if there are any) or an empty dir in `/var/run/kubevirt/container-disks`
 * Further this mount is `HostToContainer` mount propagation
 * Further we have `/var/run/kubevirt` mounted with `Bidirectional` mount propagation to virt-launcher
 * `virt-handler` does a mount from the host perspective (to see the container disk mounts) to `/var/run/kubevirt/container-disks/<uid>/<disk>`.
 * Now `virt-handler` and `virt-launcher` both see the disks in their namespaces, because of the mount propagation
 * `virt-handler` can now anytime unmount based on the watchdog file, by just checking  for mounts in the container-disks folder of the VMI. No pid is needed anymore at this stage.

Reply:
oh man. this makes my brain hurt. bottom line is it looks like it works correctly, which is all I care about.